### PR TITLE
[system-command-line] implements tab completion

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem

https://github.com/dotnet/templating/issues/3814

### Solution
Implements tab completion.

Additionally adds dotnet new3 complete sub command which prints the tab completion.
This allows to set up shells to do tab completion in the same way dotnet CLI does.

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                            TemplateCommand command = new TemplateCommand(this, environmentSettings, templatePackageManager, templateGroup, template);
+                            Parser parser = ParserFactory.CreateTemplateParser(command);
+                            ParseResult templateParseResult = parser.Parse(args.RemainingArguments ?? Array.Empty<string>());
+
+                            //TODO:
+                            // command.GetSuggestions does not work for arguments
+                            // parseResult.GetSuggestions does not work with textToMatch, only position
+                            // discuss options here
+                            //foreach (string? suggestion in command.GetSuggestions(templateParseResult, textToMatch))
+                            foreach (string? suggestion in templateParseResult.GetSuggestions())
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs:43]: there is a dilemma here:
- `command.GetSuggestions(templateParseResult, textToMatch)` doesn't work for argument completion
- `templateParseResult.GetSuggestions()` only completes end of the line. 

We need mixture of those:
- the method should determine the symbol to complete (like `ParseResult.GetSuggestions`)
- the method should consider `textToMatch` like (`Command.GetSuggestions`)

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        protected internal override IEnumerable<string> GetSuggestions(ParseResult parseResult, IEngineEnvironmentSettings environmentSettings, string? textToMatch)
+        {
+            InstantiateCommandArgs args = ParseContext(parseResult);
+
+            using TemplatePackageManager templatePackageManager = new TemplatePackageManager(environmentSettings);
+            HostSpecificDataLoader? hostSpecificDataLoader = new HostSpecificDataLoader(environmentSettings);
+
+            //TODO: consider new API to get templates only from cache (non async)
+            IReadOnlyList<ITemplateInfo> templates =
+                Task.Run(async () => await templatePackageManager.GetTemplatesAsync(default).ConfigureAwait(false)).GetAwaiter().GetResult();
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs:66]: to be extracted to new issue

File: src/dotnet-new3/CompleteCommand.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                    .Parse(input);
+
+                DefaultTemplateEngineHost host = HostFactory.CreateHost(preParseResult.GetValueForOption(New3Command.DebugDisableBuiltInTemplatesOption));
+                ITelemetryLogger telemetryLogger = new TelemetryLogger(null, preParseResult.GetValueForOption(New3Command.DebugEmitTelemetryOption));
+                string[] remainingArgs = preParseResult.GetValueForArgument(New3Command.RemainingTokensArgument) ?? Array.Empty<string>();
+
+                Command newCommand = NewCommandFactory.Create(new3Command.Name, host, telemetryLogger, new NewCommandCallbacks());
+
+                //TODO: discuss rawInput requirement
+                ParseResult newCommandResult = ParserFactory.CreateParser(newCommand).Parse(remainingArgs, rawInput: input);
```
Comment (vlada-shubina) [inline src/dotnet-new3/CompleteCommand.cs:50]: tab completion doesn't work correctly unless `rawInput` is passed, why?

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new uninstall ");
+            var result = parseResult.GetSuggestions().ToArray();
+
+            return Verifier.Verify(result, _verifySettings.Settings);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact (Skip = "https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Parsing/ParseResultExtensions.cs#L285-L289; the tab completion also contains the results from parent command")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs:60]: this test doesn't work now, the result also contain the result from base command, due to https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Parsing/ParseResultExtensions.cs#L285-L289

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new create ");
+            var result = parseResult.GetSuggestions().ToArray();
+
+            return Verifier.Verify(result, _verifySettings.Settings);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact(Skip = "https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Parsing/ParseResultExtensions.cs#L285-L289; the tab completion also contains the results from parent command")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs:110]: this test doesn't work now, the result also contain the result from base command, due to https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Parsing/ParseResultExtensions.cs#L285-L289

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new console --framework net5.0 --l");
+            var suggestions = parseResult.GetSuggestions().ToArray();
+
+            Assert.Contains("--langVersion", suggestions);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact(Skip = "not valid behavior for parser, should suggest --nuget-source")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:32]: not expected behavior.
`--interactive` is a boolean `Option`, and parser suggest `True` and `False`, which are not absolutely needed.
Excepted that parser will suggest next option instead (`--add-source`, `--nuget-source`)

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new install --nuget-source ");
+            var result = parseResult.GetSuggestions().ToArray();
+
+            Assert.Empty(result);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact(Skip = "not valid behavior for parser, should suggest --interactive probably")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:58]: not expected behavior.
`--nuget-source` is `Option` that accepts multiple args, and one arg is already given.
Excepted that parser will suggest next option instead (`--interactive`), now nothing is returned

File: test/dotnet-new3.UnitTests/DotnetNew3CompleteTests.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            commandResult
+                .Should()
+                .ExitWith(0)
+                .And.NotHaveStdErr();
+
+            Approvals.Verify(commandResult.StdOut);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact(Skip = "not working")]
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/DotnetNew3CompleteTests.cs:42]: https://github.com/dotnet/command-line-api/blob/d6046b4cbe51af1dfcf862fb64d844a8ab1bdb76/src/System.CommandLine/Parsing/ParseResultExtensions.cs#L274 invalidly determines symbol to complete here.
It determines it as `--language` option, which is already complete and at different position.

Likely this condition is wrong https://github.com/dotnet/command-line-api/blob/d6046b4cbe51af1dfcf862fb64d844a8ab1bdb76/src/System.CommandLine/Parsing/ParseResultExtensions.cs#L353-L356; it doesn't take into account if option has complete args set if position is specified.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+            var parseResult = instantiateCommand.Parse($" new foo --testChoice v");
+            var args = new InstantiateCommandArgs(instantiateCommand, parseResult);
+
+            var result = instantiateCommand.GetSuggestions(args, templateGroups, settings, packageManager, "");
+
+            Assert.Equal(new[] { "val1", "val2" }, result);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact (Skip = "not working, text to match is not considered")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:118]: ref1

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new install --nuget-source me");
+            var result = parseResult.GetSuggestions().ToArray();
+
+            Assert.Contains("--interactive", result);
+        }
+
+        [Fact]
+        public void CanCompleteChoice_FromSingleTemplate()
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:72]: ref2

**Comment by vlada-shubina** [comment]
@baronfel this branch can be used for tab completion testing.
Approach is same as in dotnet: https://docs.microsoft.com/en-us/dotnet/core/tools/enable-tab-autocomplete

Shell settings:

Powershell:
```
 # PowerShell parameter completion shim for the dotnet new3
 Register-ArgumentCompleter -Native -CommandName new3 -ScriptBlock {
     param($commandName, $wordToComplete, $cursorPosition)
         new3 complete --position $cursorPosition "$wordToComplete" | ForEach-Object {
            [System.Management.Automation.CompletionResult]::new($_, $_, 'ParameterValue', $_)
         }
 }
```
Note: `new3` should be added to PATH and point to dotnet-new3.exe

Bash:
```
alias new3=<path to dotnet-new3.exe>

_dotnet_new3_bash_complete()
{
  local word=${COMP_WORDS[COMP_CWORD]}

  local completions
  completions="$(new3 complete --position "${COMP_POINT}" "${COMP_LINE}" 2>/dev/null)"
  if [ $? -ne 0 ]; then
    completions=""
  fi

  COMPREPLY=( $(compgen -W "$completions" -- "$word") )
}

complete -f -F _dotnet_new3_bash_complete new3
```

There are known issues, specified in PR comments, there may be more of them :) 
If in doubt whether it is a shell issue or complete issue use `new3 complete "your args" --position <position of cursor>`, position is optional and if not specified means end of args line.

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs:50]: Correct way is to use `ParseResult.GetSuggestions`.  There should an API allowing to get position and raw input.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs (hunk 4 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs:60]: https://github.com/dotnet/command-line-api/issues/1494

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs (hunk 6 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:48]: https://github.com/dotnet/command-line-api/issues/1495

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs (hunk 7 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:74]: https://github.com/dotnet/command-line-api/issues/1495

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new console --framework net5.0 --l");
+            var suggestions = parseResult.GetSuggestions().ToArray();
+
+            Assert.Contains("--langVersion", suggestions);
+        }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+        [Fact (Skip = "doesn't work at the moment; it matches with legacy --language option which cannot be completed; to discuss how to avoid that")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:32]: https://github.com/dotnet/command-line-api/issues/1496

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs:50]: https://github.com/dotnet/command-line-api/issues/1497

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs (hunk 2 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs:73]: https://github.com/dotnet/templating/issues/2191

File: src/Microsoft.TemplateEngine.Cli/Commands/ParserFactory.cs [hunk 12]
```diff
@@ -11,23 +11,20 @@ namespace Microsoft.TemplateEngine.Cli.Commands
...
-            .UseHelp()
-            .UseParseErrorReporting()
-            .UseParseDirective()
-            .UseSuggestDirective()
-            .DisablePosixBundling()
-            .Build();
-        }
+            var builder = new CommandLineBuilder(command)
+                .UseParseErrorReporting()
+                //TODO: decide if it's needed to implement it; and implement if needed
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/ParserFactory.cs:18]: https://github.com/dotnet/templating/issues/2191

File: src/dotnet-new3/ParserFactory.cs [hunk 13]
```diff
@@ -17,8 +16,9 @@ internal static Parser CreateParser(Command command, bool disableHelp = false)
             var builder = new CommandLineBuilder(command)
             //.UseExceptionHandler(ExceptionHandler)
             //.UseLocalizationResources(new CommandLineValidationMessages())
-            .UseParseDirective()
-            .UseSuggestDirective()
+            //TODO: decide if it's needed to implement it; and implement if needed
```
Comment (vlada-shubina) [inline src/dotnet-new3/ParserFactory.cs:19]: https://github.com/dotnet/templating/issues/2191

File: test/Microsoft.TemplateEngine.Cli.UnitTests/Microsoft.TemplateEngine.Cli.UnitTests.csproj [hunk 14]
```diff
@@ -10,6 +10,8 @@
 
   <ItemGroup>
     <PackageReference Include="FakeItEasy" />
+    <PackageReference Include="Verify.XUnit" Version="14.2.0" />
```
Comment (DavidKarlas) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/Microsoft.TemplateEngine.Cli.UnitTests.csproj:13]: I assume this is already compatible with SourceBuild and we won't have problems with this, right?

**Review by DavidKarlas — APPROVED** [review]

File: src/dotnet-new3/CompleteCommand.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        public int Run(ParseResult result)
+        {
+            try
+            {
+                var input = result.GetValueForArgument(PathArgument) ?? string.Empty;
+                var position = result.GetValueForOption(PositionOption);
+
+                if (position > input.Length)
+                {
+                    input += " ";
```
Comment (bekir-ozturk) [inline src/dotnet-new3/CompleteCommand.cs:35]: Why is this needed? Maybe comment in the code for the next dev?

File: src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs [hunk 16]
```diff
@@ -92,7 +92,7 @@ protected IEngineEnvironmentSettings CreateEnvironmentSettings(GlobalArgs args,
             };
 
             ParseResult reparseResult = ParserFactory
-                .CreateParser(helperCommand)
+                .CreateParser(helperCommand, disableHelp: true)
```
Comment (bekir-ozturk) [inline src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs:95]: Nit: A comment to explain why we disable help could be useful.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        [Fact]
+        public Task RootCommand_GetAllSuggestions()
+        {
+            ITemplateEngineHost host = TestHost.GetVirtualHost(additionalComponents: BuiltInTemplatePackagesProviderFactory.GetComponents(includeTestTemplates: false));
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new ");
+            var result = parseResult.GetSuggestions().ToArray();
+
+            return Verifier.Verify(result, _verifySettings.Settings);
```
Comment (bekir-ozturk) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.Approval.cs:32]: Wow, this looks easy to use :eyes:

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs (hunk 11 above)
Comment (bekir-ozturk) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:32]: Should we have matching work items in our repo to track these?

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+        {
+            ITemplateEngineHost host = TestHost.GetVirtualHost(additionalComponents: BuiltInTemplatePackagesProviderFactory.GetComponents(includeTestTemplates: false));
+            var myCommand = NewCommandFactory.Create("new", host, new TelemetryLogger(null, false), new NewCommandCallbacks());
+
+            var parseResult = myCommand.Parse("new console --language ");
+            var suggestions = parseResult.GetSuggestions().ToArray();
+
+            Assert.Contains("C#", suggestions);
+            Assert.Contains("F#", suggestions);
+            Assert.Contains("VB", suggestions);
```
Comment (bekir-ozturk) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:44]: Maybe also `Assert.DoesNotContain("net5.0", suggestions);`

Comment (bekir-ozturk) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:44]: Similar cases for other tests may be added.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs (hunk 6 above)
Comment (bekir-ozturk) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:48]: Maybe we shouldn't make `--interactive` a boolean. We can just define it as `new Option("--interactive")` instead of `new Option<bool>("--interactive")`.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/Microsoft.TemplateEngine.Cli.UnitTests.csproj (hunk 14 above)
Comment (bekir-ozturk) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/Microsoft.TemplateEngine.Cli.UnitTests.csproj:13]: Tests are excluded from source build, best to my knowledge.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs (hunk 11 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:32]: I will do a separate task which tracks all disabled tests end of the week

File: test/Microsoft.TemplateEngine.Cli.UnitTests/Microsoft.TemplateEngine.Cli.UnitTests.csproj (hunk 14 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/Microsoft.TemplateEngine.Cli.UnitTests.csproj:13]: tests are not enabled in source build

File: src/dotnet-new3/CompleteCommand.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/dotnet-new3/CompleteCommand.cs:50]: no failure in validation doesn't not prevent tab completion. In fact when tab completion is called, parse result might have some errors, due to missing parameters.

File: src/dotnet-new3/CompleteCommand.cs (hunk 15 above)
Comment (vlada-shubina) [reply src/dotnet-new3/CompleteCommand.cs:35]: I can do only assumption here, probably shells are cutting trail spaces. This is copy of https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/commands/dotnet-complete/CompleteCommand.cs, I'll mention it in the code. When inserted to SDK, this command will not be used.

File: src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs (hunk 16 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs:95]: we need to pass `-h` through, if `-h` is given it would be intercepted on the first parse, hence disabling the help.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs (hunk 18 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/ParserTests/TabCompletionTests.cs:44]: added count checks
